packages/platform-core/src/object_accumulator.ts

```diff
@@ -0,0 +1,49 @@
+import _ from 'lodash';
+
+/**
+ * This error is thrown when there's a collision in
+ */
+export class ObjectAccumulatorPropertyAlreadyExistsError extends Error {
+  /**
+   * Creates property already exists error.
+   */
+  constructor(
+    readonly key: string,
+    readonly existingValue: unknown,
+    readonly incomingValue: unknown
+  ) {
+    super(`Property ${key} already exists`);
+  }
+}
+
+/**
+ * A class that can accumulate (squash merge) objects into single instance.
+ */
+export class ObjectAccumulator<T> {
+  /**
+   * creates object accumulator.
+   */
+  constructor(private readonly accumulator: Partial<T>) {}
+
+  accumulate = (part: Partial<T>): ObjectAccumulator<T> => {
+    _.mergeWith(this.accumulator, part, (existingValue, incomingValue, key) => {
+      if (_.isArray(existingValue)) {
+        return existingValue.concat(incomingValue);
+      }
+      if (existingValue && !_.isObject(existingValue)) {
```

Comment:
can `typeof existingValue !== 'object'` work here?

Reply:
seems to be working.